thoth-app/src/component/root.rs

```diff
@@ -1,77 +1,173 @@
 
-pub struct RootComponent {}
+pub struct RootComponent {
+    current_route: Option<AppRoute>,
+    account_service: AccountService,
+    current_user: Option<AccountDetails>,
+    current_user_response: Callback<Result<AccountDetails, AccountError>>,
+    current_user_task: Option<FetchTask>,
+    renew_token_task: Option<FetchTask>,
+    renew_token_response: Callback<Result<AccountDetails, AccountError>>,
+    _router_agent: Box<dyn Bridge<RouteAgent>>,
+    session_timer_agent: SessionTimerDispatcher,
+    link: ComponentLink<Self>,
+}
+
+pub enum Msg {
+    FetchCurrentUser,
+    CurrentUserResponse(Result<AccountDetails, AccountError>),
+    RenewToken,
+    RenewTokenResponse(Result<AccountDetails, AccountError>),
+    Route(Route),
+    Login(AccountDetails),
+    Logout,
+}
 
 impl Component for RootComponent {
-    type Message = ();
+    type Message = Msg;
     type Properties = ();
 
-    fn create(_: Self::Properties, _link: ComponentLink<Self>) -> Self {
-        RootComponent {}
+    fn create(_: Self::Properties, link: ComponentLink<Self>) -> Self {
+        let session_timer_agent = SessionTimerAgent::dispatcher();
+        let _router_agent = RouteAgent::bridge(link.callback(Msg::Route));
+        let route_service: RouteService = RouteService::new();
+        let route = route_service.get_route();
+
+        RootComponent {
+            current_route: AppRoute::switch(route),
+            account_service: AccountService::new(),
+            current_user: Default::default(),
+            current_user_response: link.callback(Msg::CurrentUserResponse),
+            current_user_task: Default::default(),
+            renew_token_task: Default::default(),
+            renew_token_response: link.callback(Msg::RenewTokenResponse),
+            _router_agent,
+            session_timer_agent,
+            link,
+        }
     }
 
-    fn change(&mut self, _props: Self::Properties) -> ShouldRender {
-        false
+    fn rendered(&mut self, first_render: bool) {
+        if first_render && self.account_service.is_loggedin() {
+            self.link.send_message(Msg::FetchCurrentUser);
+        }
     }
 
-    fn update(&mut self, _msg: Self::Message) -> ShouldRender {
+    fn update(&mut self, msg: Self::Message) -> ShouldRender {
+        match msg {
+            Msg::FetchCurrentUser => {
+                let task = self
+                    .account_service
+                    .account_details(self.current_user_response.clone());
+                self.current_user_task = Some(task);
+            }
+            Msg::RenewToken => {
+                let task = self
+                    .account_service
+                    .renew_token(self.renew_token_response.clone());
+                self.renew_token_task = Some(task);
+            }
+            Msg::CurrentUserResponse(Ok(account_details)) => {
+                self.link.send_message(Msg::Login(account_details));
+                self.current_user_task = None;
```

Comment:
Could this (and the one in the `Msg::RenewTokenResponse` case) be moved into the `Msg::Logout` case?

Reply:
Not quite, otherwise we'd logout the user just after we fetch their details

Reply:
Oops, meant to put this comment on the `CurrentUserResponse(Err)` case below, not this one...

Reply:
To be clear, I meant the `self.current_user_task = None;` line only - it looked more obvious when the review was in draft!

Reply:
Oh I see. In any case I want to separate the actions, unsetting `current_user_task` should be done when dealing with the response, rather than at logout